include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence, typename Handler>
+    auto async_read_some(MutableBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        if (decrypted_buffer.size() > 0) {
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            next_layer().get_io_service().dispatch([
+                consumed, handler = std::move(init.completion_handler)
+            ]() mutable {
+                beast::error_code error{};
+                handler(error, consumed);
+            });
+        } else {
+            async_read_some_op<MutableBufferSequence, Handler> operation{
+                *this, sequence, std::move(init.completion_handler)};
+            next_layer().get_io_service().dispatch([operation = std::move(
+                                                        operation)]() mutable {
+                beast::error_code error{};
+                operation(error, 0);
+            });
+        }
+        return init.result.get();
+    }
+
+    template <typename ConstBufferSequence, typename Handler>
+    auto async_write_some(ConstBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        async_write_some_op<ConstBufferSequence, Handler> operation{
+            *this, sequence, std::move(init.completion_handler)};
+        next_layer().get_io_service().dispatch([operation = std::move(
+                                                    operation)]() mutable {
+            beast::error_code error{};
+            operation(error, 0);
+        });
+        return init.result.get();
+    }
+
+    void set_secrets(const crypto::aead_secrets& secrets) {
+        this->secrets = secrets;
+    }
+
+private:
+    template <typename ConstBufferSequence, typename Handler>
+    struct async_write_some_op : public asio::coroutine {
+        async_write_some_op(secure_socket& socket,
+                            const ConstBufferSequence& sequence,
+                            Handler&& handler)
+            : socket{socket},
+              sequence{sequence},
+              handler{std::forward<Handler>(handler)} {}
+
+        void operator()(beast::error_code error, size_t n) {
+            CORO_REENTER(*this) {
+                static_cast<void>(n);
+                socket.encrypt_pdu(sequence, pdu);
+                CORO_YIELD asio::async_write(socket.next_layer(),
+                                             asio::buffer(pdu), *this);
+                handler(error, asio::buffer_size(sequence));
+            }
+        }
+
+        secure_socket& socket;
+        ConstBufferSequence sequence;
+        Handler handler;
+        std::vector<gsl::byte> pdu;
```

Comment:
Mind you that every time you pass `*this`, Asio is making a copy of your handler (which is a coroutine). That's fine for the `socket`, `sequence`, and `handler` members but for the `std::vector` you are paying for an unnecessary dynamic allocation. This kills performance. Usually people put expensive to copy objects in a `std::shared_ptr`.

You cannot instead pass `std::move(*this)` (as you can with normal completion handlers) because you are using coroutines and that would break it.